passwdmancli.1
.\"Copyright (c) 2013, 2014, Oskar Skog <oskar.skog.finland@gmail.com>
.\"All rights reserved.
.\"
.\"Redistribution and use in source and binary forms, with or without
.\"modification, are permitted provided that the following conditions are met:
.\"
.\"1.  Redistributions of source code must retain the above copyright notice,
.\"    this list of conditions and the following disclaimer.
.\"
.\"2.  Redistributions in binary form must reproduce the above copyright notice,
.\"    this list of conditions and the following disclaimer in the documentation
.\"    and/or other materials provided with the distribution.
.\"
.\"THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\"AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\"IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\"ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
.\"LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\"CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\"SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\"INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\"CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\"ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\"POSSIBILITY OF SUCH DAMAGE.
.Dd Jan 27, 2014
.Dt PASSWDMANCLI 1
.Os
.Sh NAME
.Nm passwdmancli
.Nd CLI for storing, generating and updating passwords and honeypots.
.Sh SYNOPSIS
.Nm
.Op Fl i
.Op Fl -interactive
.Op Fl B
.Op Fl -single-backup
.Op Fl U
.Op Fl -utf-8
.Op Ar infile
.Sh DESCRIPTION
.Ar infile
is the file
.Nm \"passwdmancli
will read its commands from, if
.Ar infile
is omitted
.Nm \"passwdmancli
will read from standard input.
.Fl i
or
.Fl -interactive
must be specified if you want the commands
.Ic undo
and
.Ic redo ,
or if you want
.Nm
to stay alive after an error.
.Pp
If you specify the option
.Fl B
or
.Fl -single-backup ,
every change that has been done, will be removed by TWO undos.
.Pp
You can force
.Nm
to use
.Qq utf-8
as encoding by specifying the option
.Fl -utf-8 .
WARNING: It hasn't been tested with Python 3.x in a non-UTF-8 environment.
.Pp
.Nm
reads its input as colon separated values.
Quotes are either forbidden or optional (look at 
.Fn unquote
in
.Xr passwdmanapi 3
).
Integers are created with the python built-in
.Fn int .
All commands are listed in
.Sx COMMANDS .
The first command you really should use is
.Qq esc:on .
If the line begins with a number-sign, the line will be ignored.
If the line begins with a pipe-sign, it will write the rest of the line to
standard output.
.Pp
All outputs that are prefixed with an exclamation mark are error messages.
All outputs that are prefixed with a number sign are messages for a human
user.
All other outputs are prefixed with their command (not their arguments),
using dots instead of colons, followed by a question mark.
.Pp
Beware of whitespace.
The whitespace comes after colons and/or question marks and before quotes.
.Pp
The passwords have meta-data.
They have a minimum required length and a maximum allowed length.
They also have a type which is one of:
.Bl -diag
.It "10"
The password may only use digits.
.It "64"
The password can use big letters, small letters, digits, underscores and
exclamation marks.
.It "human"
The password is human generated.
.El
.Pp
The
.Qq honeypots
are weak passwords intended to be used as traps.
It was a poor choice of name, but it grew legs and glued its feet to the
ground.
.Pp
WARNING: If a command accepts a
.Va name
and
.Va index
argument, any string of digits will be treated as an index.
.Sh COMMANDS
.Bl -tag
.It "bells"
Show message about the bells and whistles.
.It "bugs"
Show message about wasps and mosquitoes.
.It "esc:off"
Disable really simple escapes for quoted strings.
.It "esc:on"
Enable really simple escapes for quoted strings.
They are only useful for escaping quotes.
.It "hell"
Synonym for
.Ic "verbose:on" .
.It "help"
Help message.
.It "honeypot:add:'value'"
Add a new honeypot with the value
.Va value .
.It "honeypot:help"
Help message.
.It "honeypot:list"
List all honeypots.
Returns multiple lines of:
.Qq honeypot.list? ,
the index, a colon and the value surrounded by single quotes.
.It "honeypot:pick"
Pick a honeypot.
Returns a line of:
.Qq honeypot.pick?
followed by the honeypot surrounded by single quotes.
.It "honeypot:pick:n:'sep'"
Pick
.Va n
honeypots and separate them with
.Va sep .
Returns a line of:
.Qq honeypot.pick?
followed by the separator-separated honeypots surrounded by single quotes.
.It "honeypot:pickl:n"
Pick
.Va n
honey-pots, escape and quote (single quotes).
Returns a line of
.Qq honeypot.pickl?
followed by the quoted honey-pots separated with commas.
.It "honeypot:remove:'value'"
Remove the honeypot with the value
.Va value .
.It "honeypot:remove:index"
Remove the honeypot at
.Va index .
WARNING: The indexes of the following honeypots will be decreased by one!
.It "passwd:add:type:minlength:maxlength:'name'"
Add a new password for
.Va name ,
with at least
.Va minlength
letters, but no more than
.Va maxlength .
If
.Va type
is
.Qq 10 ,
the password will only contain digits.
If
.Va type
is
.Qq 64 ,
the password will contain big letters, small letters, digits, underscores and
exclamation marks.
It will not give any output.
.It "passwd:add_human:'name'"
Add a human generated password for
.Va name .
It will read the password on a new line.
It will not give any output.
.It "passwd:get:'name'"
Returns a line of:
.Qq passwd.get?
followed by the value of the password for
.Va name
surrounded by single quotes.
.It "passwd:get:index"
Returns a line of:
.Qq passwd.get?
followed by the value of the password at
.Va index
surrounded by single quotes.
.It "passwd:help"
Help message.
.It "passwd:list"
List all passwords.
Returns multiple lines of:
.Qq passwd.list? ,
the index, a colon and the name surrounded by single quotes.
.It "passwd:meta:'name'"
Get the meta-data for the password for
.Va name .
Returns a line of:
.Qq passwd.meta? ,
the type surrounded by single quotes, a colon, the minimum length for the
password surrounded by single quotes, a colon and the maximum length for the
password surrounded by single quotes.
.It "passwd:meta:index"
Get the meta-data for the password at
.Va index .
Returns a line of:
.Qq passwd.meta? ,
the type surrounded by single quotes, a colon, the minimum length for the
password surrounded by single quotes, a colon and the maximum length for the
password surrounded by single quotes.
.It "passwd:remove:'name'"
Remove the password for
.Va name .
.It "passwd:remove:index"
Remove the password at
.Va index .
WARNING: The indexes of the following passwords will be decreased by one!
.It "passwd:update:'name'"
Update the password for
.Va name
using its meta-data.
Returns a line of:
.Qq passwd.update? ,
the old password, a colon and the new password.
.It "passwd:update:index"
Update the password at
.Va index
using its meta-data.
Returns a line of:
.Qq passwd.update? ,
the old password, a colon and the new password.
.It "passwd:update_meta:type:min:max:'name'"
Do the same thing as
.Ic "passwd:update:'name'" ,
but use
.Va type ,
.Va min ,
and
.Va max
as meta-data, instead of the passwords own (which might not even exist).
.It "passwd:update_meta:type:min:max:index"
Do the same thing as
.Ic "passwd:update:index" ,
but use
.Va type ,
.Va min ,
and
.Va max
as meta-data, instead of the passwords own (which might not even exist).
.It "pb:on"
Enable progress-counters.
start is
.Qq Number-sign, tab, zero, percent.
Progress is
.Qq Enough backspaces, integer decimal digits, percent.
Stop is
.Qq Newline.
.It "pb:off"
Disable progress-counters.
.It "redo"
Opposite of
.Ic undo .
.It "undo"
CLI undo button; restore from latest auto-generated backup.
.It "verbose:on"
Enable verbose mode; release mosquitoes from hell that will be noisy and give
you a headache.
.It "verbose:off"
Disable verbose mode; kill the mosquitoes.
.It "whistles"
Show message about the bells and whistles.
.El
.Sh FILES
.Pa ~/.passwdman/passwords
is a XML file containing all passwords.
.Pp
.Pa ~/.passwdman/honeypots
is a XML file containing all honeypots.
.Pp
.Pa ~/.passwdman/undoable
is a directory full of backups.
.Pp
.Pa ~/.passwdman/redoable
is a directory full of backups.
.Sh DIAGNOSIS
All error messages are sent to standard output and are prefixed with an
exclamation mark.
.Bl -diag
.It "!syntax error"
Bad speling or incorrect usage.
The next line of output will be the offending input prefixed with an
exclamation mark.
.It "!no meta"
The command requires meta-data, but the password doesn't have any.
.It "!not found"
Not found or index out of range
.El
.Sh SEE ALSO
.Xr passwdmanapi 3 ,
.Xr random 4 ,
.Xr urandom 4 ,
.Xr passwdmanfiles 5
.Sh BUGS
Mosquitoes:
.Bl -bullet
.It
If verbose mode is enabled the message
.Qq #READY
is written too many times.
.El
.Sh SECURITY
.Nm
uses
.Xr random 4
or
.Xr urandom 4
for everything.
.Sh AUTHOR
Written by
.An Oskar Skog Aq oskar.skog.finland@gmail.com .
.Pp
Please send patches, questions, bug reports and wish-lists.

NEWS
0.4.15 Fixed really bad bug.
%
0.4.14  The 0's are *slushed*.
    A complete restart is in progress.
    1.x         No stupid misfeature a.k.a fake-passwords.
    1.x         10/64/human will be replaced by a generic customizable
                randomizer.
    1.x         os.urandom()
    1.x         Unicode since day 0. (With gettext friendly side-effect.)
    1.x         Non-stupid CLI.
    1.x         No progress-bars, they were useless.
%
0.4 is here.
    *   Easier to read code. Lots of code -> data.
    *   Rogue-like interface. RLI
    *   Progress bars (rogue-like interface).
    *   Works with Python 2 and 3.
    *   Works on FreeBSD.
    *   The RLI and the CLI uses the locale's encoding. 
    *   passwd.add can randomize the password itself.
%
0.3 is here.
    *   HTML docs validates.
    *   Edit history buttons.
    *   Clear clipboard button.
    *   in configure.2: getmoduledir() picks newest 2.*.
    *   Single change mode for the CLI. 2 undos undoes everything.
    *   passwd.add_nometa() doesn't add meta-data to the file.
    *   In configure: --update works without argument.
%
0.2 is here.
    *   All Docbook XML generates good HTML, but passwdmanapi.xml requires
        --skip-validation.
    *   More scrollbars.
    *   A "Make backup"-button.
    *   PDF and postscript documentation (papersize = A4).
    *   Better looking data files.
    *   Better makefile.
    *   Useful CLI. You need to specify the -i, --interactive to enable undo,
        redo and survive errors.
%


passwdmancli.py
#!/usr/bin/python
# -*- encoding: utf-8 -*-

copywrong = """#Copyright (c) 2013, Oskar Skog <oskar.skog.finland@gmail.com>
#All rights reserved.
#
#Redistribution and use in source and binary forms, with or without
#modification, are permitted provided that the following conditions are met:
#
#1.  Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#2.  Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
#THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
#AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
#IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
#ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
#LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
#CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
#SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
#INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
#CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
#ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#POSSIBILITY OF SUCH DAMAGE.
"""

import passwdmanapi as api
import sys
import logging
import time
import os
import os.path
import getopt

def se():
    """se() - syntax error
    Show error message containing theline.
    """
    sow("!syntax error, type 'help'\n")
    sow('!')
    sow(theline)        #str.format() doesn't like Unicode.
    sow("\n")
    if not interactive:
        sys.exit(1)

def err(x):
    """err(x) - error messages
    "!{}\n".format(x)
    Program dies if not in interactive mode.
    """
    sow("!")
    sow(x)
    sow("\n")
    if not interactive:
        sys.exit(1)

def sow(x):             #Do the Unicode-stuff here.
    """Write to stdout. It encodes the string as UTF-8."""
    sys.stdout.write(x.encode(encoding="utf-8"))         #It is used a lot.

def siw():
    """Read from infile. It decodes the UTF-8 input."""
    global infile
    return infile.readline().decode(encoding="utf-8")

def v(x):       #Verbose.
    """def v(x):
    global verbose
    if verbose:
        sow("#" + x + "\n")
"""
    if verbose:
        sow("#" + x + "\n")     #Make the output comments

def show_help():
    """#HELP MESSAGE
#passwdmancli uses  colon separated fields  as input. Everything inside angle
#brackets < >,  is intended to be expanded in your brain.
#It tries to be more grepable by making the first letter in error messages be
#an exclamation  mark and  the first  letter in human-friendly  messages be a
#hash/number sign.  Multi-line  outputs are  terminated by a "#READY"
#message. Unless otherwise noted, quotes  around angle  brackets are optional
#and may be either single quotes or double quotes. If the first letter in its
#input is a '#', it will ignore the line. If the first letter in it input is
# '|', it prints out the rest of the line
#It allows ANY crap to surround the commands in the fields.
#It's outputs are prepended with `<command> | tr : .` + '?'
#The error messages are hopefully obvious enough.
#passwd:help            Show help for passwd:*
#honeypot:help          Show help for honeypot:*
#verbose:on            Verbose output, the command 'hell' does the same thing.
#verbose:off            Brief output.
#undo                   Undo last change to passwd or honeypot, there is also
#help                   Obvious.                                'redo'.
#exit                   Obvious.
#bugs                   Bugs.
#whistles               Bells and whistles.
"""
    sow(show_help.__doc__)
def passwd_help():      #22 lines + ('#READY') + new line for input
    """#HELP MESSAGE
#Do you remember the angle brackets and the quotes? If not: type 'help'
#passwd:add:<type>:<min>:<max>:"<name>"         Add password, for <name> with
#                       a minimal length of <min> and maximal length of <max>
#                       <type> == 10 = use digits only, <type> == 64 = use
#                       big and small letters, digits, '!' and '_'.
#passwd:add_human:"<name>"                      Add human generated password
#                       for <name>. The password is entered on a new line.
#passwd:remove:<index>                          Remove, by index.
#passwd:remove:"<name>"                         Remove, by name.
#passwd:get:<index>     passwd:get:"<name>"     Get password.
#passwd:meta:<index>                            Get meta-data, which is...
#passwd:meta:"<name>"                           ...useful when updating.
#passwd:remove:<index>          passwd:remove:"<name>"       remove password
#passwd:list                              list all passwords (index and name)
#passwd:update:"<name>"         passwd:update:<index>   update password, will
#                       print out the new and the old.
#passwd:update_meta:<type>:<min>:<max>:"<name>"      Update the password,...
#passwd:update_meta:<type>:<min>:<max>:<index>       ...but use new meta-data
#                                       (the <type>, <min> and <max>)
#               passwd:update      reuses the existing meta-data
#
"""
    sow(passwd_help.__doc__)

def honeypot_help():
    """#HELP MESSAGE
#Do you remember the angle brackets and the quotes? If not: type 'help'
#The honey pots are human generated  passwords, whose only purposes are to be
#used as traps.
#honeypot:add:"<value>"                 Add a new honey pot.
#honeypot:pick                          Pick a random honeypot.
#honeypot:pick:<n>:"<sep>"              Pick <n> random honeypots use <sep>.
#                                       A separator between them.
#honeypot:remove:"<value>"              Remove.
#honeypot:list                          List.
#
#
#
#
#
#
#
#
#
#
#
#
"""
    sow(honeypot_help.__doc__)
   
def bells_and_whistles():
    """#BELLS AND WHISTLES
#       *       Comments:       It ignores input-lines that begins with a '#'
#
#       *       Pipe-through:   If it's input begins with a '|', the rest of
#               the line is written to stdout.
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
#
"""
    sow(bells_and_whistles.__doc__)

def bugs():
    """#BUGS
#If verbose mode is enabled, it will shout '#READY' to stdout way too much.
#
#
#
#It is noisy and babbles to much.
#It uses the following methods to parse it's input:
#       int()           python built-in
#       'if "keyword" in some_variable:'
#Some of these big message screens looks ugly.
#Bad speling in the error messages.
#
#
#
#
#
#
#
#
#
#
#
"""
    sow(bugs.__doc__)
    
def passwd_update(x):
    try:                        #unquote gets rid of optional quotes.
        index = p.mkindex(api.unquote(x))  #mkindex greps it's argument...
    except:                             #...in the list and returns an index.
        err("not found")
        return
    old = p[index]["value"]       #It must show the old password, because...
                              #...it is probably required to change password.
    v("This may take a while.")       #Verbose.
    try:
        p.update(index)              #Do lots of the work in the right place.
    except api.err_nometa:
        err("no meta")               #It requires the meta-data.
        return
    new = p[index]["value"]
    sow("#old:new\n")
    #old and new doesn't contain any Unicode.
    sow("passwd.update?'{}':'{}'\n".format(old, new)) #Always 64 or 10.
    return
    
def passwd_update_meta(x):              #Update with new meta-data.
    fields = x.split(':', 3)
    if len(fields) < 4: #It takes 4 arguments.
        se() #Syntax error.
        return
    name = api.unquote(fields[3])
    try:
        index = p.mkindex(name)     #Where is it?
    except:
        err("not found")
        return
    old = p[index]["value"]             #Remember old.
    v("This may take a while.")
    try:
        p.update_meta(index, fields[0], fields[1], fields[2]) #Put the work...
                                                    #...at the right place.
    except api.err_idiot:       #Incorrect usage.
        err("stupid arguments")
        return
    new = p[index]["value"]
    sow("#old:new\n")
    sow("passwd.update?'")
    sow(old)    #old can be anything.
    sow("':'{}'\n".format(new)) #new is 64 or 10.
    return
    
def passwd_add(x):
    #passwd:add:<type>:<min>:<max>:<name>
    add_args = x.split(':', 3)
    if len(add_args) < 4:
        se()
        return
    add_type = add_args[0]   #"64" or "10".
    try:
        add_min = int(add_args[1])
        add_max = int(add_args[2])
    except:     #it requires integers
        se()
        return
    add_name = api.unquote(add_args[3])
    try:
        int(add_name)          #Check if it will be treated as an integer.
        sow("#WARNING: ITS NAME WILL BE BELIEVED TO BE AN INDEX!\n")
        sow("#IF YOU REMOVE IT BY ITS NAME, BAD THINGS CAN HAPPEN!\n")
    except:
        pass
    #start testing add_type
    if "human" in add_type:
        se()
        sow("#Use 'passwd:add_human'.\n") #add_human is inside passwd_cmds().
        return
    elif "10" in add_type:
        try:            #Use digits only.
            p.add(add_name, api.get10(api.getint(add_min, add_max + 1)), 
                  "10", str(add_min), str(add_max))
        except api.err_duplicate:
            err("duplicate")
            return
        v("Added base10 password.")
        return
    elif "64" in add_type:      #WARNING: Copy-pasted from 'elif "10"...'
        #                        and hand-hacked '10'->'64'.
        try:    #A-Z a-z '_' '!'
            p.add(add_name, api.get64(api.getint(add_min, add_max + 1)), 
                  "64", str(add_min), str(add_max))
        except api.err_duplicate:
            err("duplicate")
            return
        v("Added base64 password.")
        return
    else:
        se()
        return

def passwd_cmds(x):  #Put all the passwd:* here.
    twofields = x.split(':', 1)
    a = twofields[0]
    if len(twofields) < 2:
        if "list" in a:         #list doesn't take any arguments.
            index = 0
            sow("#\t\tindex:\tname\n")   #Human readable message.
            for l in p:
                sow("passwd.list?\t{}:\t'".format(index))
                sow(l["name"])
                sow("'\n")
                index += 1
            sow("#READY\n")
            return
        elif "help" in a:       #All big messages are done by calling a...
            passwd_help()       #...function that prints out it's doc-string.
            return
        else:
            se()
            return
    else: #passwd:*:*
        b = twofields[1]
        if "add_human" in a:            #WARNING: Must be before "add".
            s = siw()
            name = api.unquote(b)
            try:
                int(name)          #Check if it will be treated as an integer.
                sow("#WARNING: ITS NAME WILL BE BELIEVED TO BE AN INDEX!\n")
                sow("#IF YOU REMOVE IT BY ITS NAME, BAD THINGS CAN HAPPEN!\n")
            except:
                pass
            try:
                p.add_nometa(name, s[:-1])    #Do the work at the right place.
            except api.err_duplicate:
                err("duplicate")
                return
            v("Added human generated password.")
            return
        elif "add" in a:                #WARNING: Must be after "add_human".
            passwd_add(b)       #Go up.
            return
        elif "get" in a:      #Get the value of a password; get the password.
            try:
                the_value = p[p.mkindex(api.unquote(b))]["value"]
            except:
                err("not found")
                return
            sow("passwd.get?'")
            sow(the_value)
            sow("'\n")
            return
        elif "update_meta" in a:        #WARNING: Must be before "update" and
                                        #"meta".
            passwd_update_meta(b)       #Go up.
            return
        elif "update" in a:             #WARNING: Must be after "update_meta".
            passwd_update(b)
            return
        elif "meta" in a:               #WARNING: Must be after "update_meta".
            try:                #Get meta-data for a password.
                metadata = p[p.mkindex(api.unquote(b))]["meta"]
                sow("#\t\ttype:\tminlength:\tmaxlength\n")      #safe
                sow("passwd.meta?\t'{}':\t'{}':\t\t'{}'\n".format(
                                                        metadata["type"],
                                                       metadata["minlength"],
                                                      metadata["maxlength"]))
            except:
                err("no meta")
            return
        elif "remove" in a:
            try:
                p.remove(api.unquote(b))
            except api.err_notfound:
                err("not found")
                return
            return
        else:
            se()
            return

def honeypot_cmds(x):           #honeypot:*
    twofields = x.split(':', 1)
    a = twofields[0]
    if len(twofields) < 2:
        if "list" in a:
            index = 0
            sow("#\t\tindex:\tname\n")
            for value in h:
                sow("honeypot.list?\t{}:\t'".format(index))
                sow(value)
                sow("'\n")
                index += 1
            sow("#READY\n")
            return
        elif "help" in a:
            honeypot_help()
            return
        elif "pick" in a:       #Pick a random honeypot.
            #pick takes either 0 or 2 arguments, look down.
            #This pick takes 0.
            v("This may take a while.")
            sow("honeypot.pick?'")
            sow(h.pick())
            sow("'\n")
            return
        else:
            se()
            return
    else:       #honeypot:*:*
        b = twofields[1]
        if "add" in a:
            name = api.unquote(b)
            try:
                int(name)          #Check if it will be treated as an integer.
                sow("#WARNING: ITS NAME WILL BE BELIEVED TO BE AN INDEX!\n")
                sow("#IF YOU REMOVE IT BY ITS NAME, BAD THINGS CAN HAPPEN!\n")
            except:
                pass
            try:
                h.add(name)
            except api.err_duplicate:
                err("duplicate")
                return
            return
        elif "pick" in a:       #pick takes either 0 or 2 arguments.
            args = b.split(':', 1)
            if len(args) < 2:   #It does allow 0, reread this function.
                se()
                return
            try:
                n = int(args[0])
            except:
                err("n must be an integer")
                return
            sep = api.unquote(args[1])
            try:                       #3rd arg, make it raise instead of log.
                sow("honeypot.pick?'")
                sow(h.pick(n, sep, False))
                sow("'\n")
            except api.err_idiot:
                err("n is too big")
                return
            return
        elif "remove" in a:
            try:
                h.remove(api.unquote(b))
            except api.err_notfound:
                err("not found")
                return
            return
        else:
            se()
            return

def main():             #Finally.
    global verbose
    global p
    global h
    global theline
    sow(copywrong)
    while True:
        v("READY")  #Show this message after every command, if verbose.
                    #It should be moved to line 666.
        temporary_variable = siw()
        if len(temporary_variable) < 1: #Not even an empty line.
            quit()
        theline = temporary_variable[:-1]       #Get rid of newline.
        if len(theline) < 1:
            continue
        if theline[0] == "#":           #Comment.
            continue
        if theline[0] == "|":           #Pipe through.
            sow(theline[1:])
            sow("\n")
            continue
        twofields = theline.split(':', 1)
        a = twofields[0]
        if len(twofields) < 2:  #Check some really short ones.
            if "exit" in a:
                quit()
            elif "help" in a:
                show_help()
            elif "undo" in a:
                if interactive:
                    try:
                        api.undo(p, h)
                    except:
                        err("cannot undo")
                        continue
                    v("Undone something.")
                else:
                    err("Not in interactive mode.")
            elif "redo" in a:
                if interactive:
                    try:
                        api.redo(p, h)
                    except:
                        err("cannot redo")
                        continue
                    v("Redone something.")
                else:
                    err("Not in interactive mode.")
            elif "bells" in a or "whistles" in a:
                bells_and_whistles()
                continue
            elif "bugs" in a:
                bugs()
                continue
            elif "hell" in a:           #Synonym for verbose:on
                verbose = True
                sow("#verbose mode enabled\n")
            else:
                se()    #syntax-error
        else: #Check some longer ones.
            #the string has been split
            b = twofields[1]
            if "passwd" in a:
                passwd_cmds(b)      #passwd:*
            elif "honeypot" in a:
                honeypot_cmds(b)    #honeypot:*
            elif "verbose" in a:
                if "on" in b:           #666
                    verbose = True
                    sow("#verbose mode enabled\n")
                elif "off" in b:
                    verbose = False
                else:
                    se()
                    continue
            else:
                se()
                continue

#initialization
p = api.passwd()
h = api.honeypot()
verbose = False
interactive = False
theline = ""
logging.basicConfig(level=logging.CRITICAL)     #STFU
if __name__ == "__main__":
    o, a = getopt.getopt(sys.argv[1:], "i", ["interactive"])
    if len(a) > 0:
        infile = open(a[0], "r")
    else:
        #infile = os.dup(sys.stdin.fileno())
        infile = sys.stdin
    for x, y in o:
        if x in ("-i", "--interactive"):
            interactive = True
    main()